### PR TITLE
br: Add retry limit of connection err in fine grained backup

### PR DESCRIPTION
**Pull request by RidRisR**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/57449

Problem Summary:

The fine grained backup may fails into infinite retry when encounter disconnect err.

### What changed and how does it work?

Add a retry limit of 3 to each store.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Try to make error in resp 3 times, will get:

Error: failed to connect to store 1 for 3 times: failed to connect to store 1: [BR:Common:ErrFailedToConnect]failed to make gRPC channels

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/backup/client.go [hunk 1]
```diff
@@ -1168,8 +1183,20 @@ func (bc *Client) fineGrainedBackup(
...
+				}
+				storeErr, ok := err.(*StoreBasedErr)
+				if !ok {
+					return errors.Trace(err)
+				}
+
+				storeID := storeErr.storeID
+				message := storeErr.message
+				maxDisconnect[storeID]++
+				if maxDisconnect[storeID] > 3 {
```
Comment (Tristan1900) [inline br/pkg/backup/client.go:1197]: nit, maybe make it a const?

File: br/pkg/backup/client.go [hunk 2]
```diff
@@ -1270,12 +1297,17 @@ func (bc *Client) handleFineGrained(
...
+
 	if err != nil {
 		if berrors.Is(err, berrors.ErrFailedToConnect) {
 			// When the leader store is died,
 			// 20s for the default max duration before the raft election timer fires.
 			logutil.CL(ctx).Warn("failed to connect to store, skipping", logutil.ShortError(err), zap.Uint64("storeID", storeID))
-			return 20000, nil
+			return 20000, &StoreBasedErr{
+				storeID: storeID,
+				message: "failed to connect to store",
```
Comment (Tristan1900) [inline br/pkg/backup/client.go:1308]: does it need message? looks like it's always using the same message

**Review by Tristan1900 — COMMENTED** [review]
LGTM, do we need a test to verify it's working as expected?

File: br/pkg/backup/client.go (hunk 2 above)
Comment (RidRisR) [reply br/pkg/backup/client.go:1308]: Done

File: br/pkg/backup/client.go (hunk 1 above)
Comment (RidRisR) [reply br/pkg/backup/client.go:1197]: A better way is reusing the `backupRetryTimes`
Done

**Comment by RidRisR** [comment]
> LGTM, do we need a test to verify it's working as expected?

Unfortunately there is no handy infrastructure for testing fine grained, and it seems not so worthy to build one for a up to deprecate function.

File: br/pkg/backup/client.go [hunk 3]
```diff
@@ -1270,12 +1295,16 @@ func (bc *Client) handleFineGrained(
...
 	lockResolver := bc.mgr.GetLockResolver()
 	client, err := bc.mgr.GetBackupClient(ctx, storeID)
+
 	if err != nil {
 		if berrors.Is(err, berrors.ErrFailedToConnect) {
 			// When the leader store is died,
 			// 20s for the default max duration before the raft election timer fires.
 			logutil.CL(ctx).Warn("failed to connect to store, skipping", logutil.ShortError(err), zap.Uint64("storeID", storeID))
-			return 20000, nil
+			return 20000, &StoreBasedErr{
```
Comment (BornChanger) [inline br/pkg/backup/client.go:1304]: better to define a const for `2000` here for better readability.

File: br/pkg/backup/client.go [hunk 4]
```diff
@@ -1314,7 +1343,10 @@ func (bc *Client) handleFineGrained(
...
```
Comment (BornChanger) [inline br/pkg/backup/client.go:1346]: ditto

**Comment by BornChanger** [comment]
> > LGTM, do we need a test to verify it's working as expected?
> 
> Unfortunately there is no handy infrastructure for testing fine grained, and it seems not so worthy to build one for a up to deprecate function.
> 
> I impl a basic manual test for it.

I have different opinion though but we need to add test case to cover fine grained case since the issue to address is one missed.

File: br/pkg/backup/client.go (hunk 4 above)
Comment (RidRisR) [reply br/pkg/backup/client.go]: done

**Comment by RidRisR** [comment]
> > > LGTM, do we need a test to verify it's working as expected?
> > 
> > 
> > Unfortunately there is no handy infrastructure for testing fine grained, and it seems not so worthy to build one for a up to deprecate function.
> > I impl a basic manual test for it.
> 
> I have different opinion though but we need to add test case to cover fine grained case since the issue to address is one missed. And those LTS versions will be kept for many years.

Maybe, but seems now we lark of infrastructure to visualize or even trigger the fine grained backup.

**Review by Leavrth — APPROVED** [review]

**Review by BornChanger — APPROVED** [review]

**Pull request merged**